### PR TITLE
Cleanup Audio background threads on shutdown

### PR DESCRIPTION
**Pull request by zzmp**
- Calls out code that behaves unexpectedly with `FIXME` labels
- Standardizes the interface of the Audio background threads
- Actually stops the Audio background threads during shutdown, to avoid use-after-free of the client's memory (fixes fogbugz 4244)

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: cd2665fc55e2dc7a39856e90ae643258e3c92555 available at: https://gist.github.com/0f5176a6da4c60612886df228214d1f7<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10269-cd2665f.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10269-cd2665f.dmg

**Comment by ZappoMan** [comment]
since this changes threads, I'd recommend the test pass include testing on an i5

File: libraries/audio-client/src/AudioClient.cpp [hunk 1]
```diff
@@ -76,42 +76,58 @@ using Mutex = std::mutex;
...
     }
 
 private:
-    AudioClient* _audioClient { nullptr };
-    Mutex _checkDevicesMutex;
-    bool _quit { false };
+    std::atomic<bool> _quit { false };
+    bool running { true };
+    std::mutex mutex;
+    std::condition_variable cv;
```
Comment (kencooke) [inline libraries/audio-client/src/AudioClient.cpp]: seems like these should be _running, _mutex, _cv per the Coding Standard.

Comment (ZappoMan) [inline libraries/audio-client/src/AudioClient.cpp]: indeed... and `_cv` should really be something like `_doneRunning` or `_quitting` or something that clearly articulates to the reader what condition you're waiting on.

**Comment by zzmp** [comment]
### QA Plan

_This should be tested on an i5 and an i7 machine, as it affects threading._

- Start up interface.
- Make sure you have a working microphone.
- Start up some local injectors (local sounds);
    ```js
    sound = SoundCache.getSound("http://xiph.org/~xiphmont/demo/1kHz-0-dither.wav");
    // wait for sound to load, or the audio will not play
    var injector = Audio.playSound(sound, { position: MyAvatar.position, loop: false, volume: 0.1 });
    ```
- While speaking in the mic and running the local sounds, quit Interface.

Interface should not crash on quit. Try this a few times.
Occasionally it will crash on `master` - you may or may not experience this.

**Comment by hifi-gustavo** [comment]
Build console output for commit: c27dd446ebfdfd2654f0993fe970b1ff437988ce available at: https://gist.github.com/0f5176a6da4c60612886df228214d1f7<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10269-c27dd44.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10269-c27dd44.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: e9c5b8600367be234126f979088fd60ffe7789a5 available at: https://gist.github.com/0f5176a6da4c60612886df228214d1f7<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10269-e9c5b86.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10269-e9c5b86.dmg

**Comment by claudioshg** [comment]
testing

**Comment by claudioshg** [comment]
### Desktop PC - Windows 10 Home - CPU 4.0GHz Intel i7 - GTX1070 - RAM 16GB - Tested with Oculus Rift

Tested: Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10269-e9c5b86.exe

---------FAIL---------

Tested in Domain: http://Sandbox
Interject Sounds used where:
- Ambisonic - Babbling Creek
- Ambisonic - Diner
- "Crickets" - Ambient Sounds

HMD Mode with Tablet: Crashed
HMD Mode with Toolbar: Closed Interface with no problem
Desktop Mode with Tablet: Crashed
Desktop Mode with Toolbar: Closed Interface with no problem

Fails:
- HMD with Tablet: I made noise into the mic as I quit the interface. All 3 ambient sounds played at the same time. The Interface crashed every time I quit and not once did a Crash Report window appear. I tested this step 6 times.

- Desktop Mode with Tablet: Same as above. The only difference is that instead of an Interface crash, my PC crashed. I received a blue screen with a message from Windows apologizing and how it was sending a crash report back to Microsoft. My PC crashed all 3 times I tested quitting while in Desktop Mode using the Tablet.

**Comment by zzmp** [comment]
It's strange that this would induce a tablet crash. I'm afraid it may just be the tablet that is crashing.
I'm talking to @claudioshg for more info.

**Comment by zzmp** [comment]
The crashes were tablet bugs. I am willing to call this QA pass, but it's also OK to wait for the tablet bugs to be resolved and retest.

**Comment by zzmp** [comment]
When #10354 is merged, I will rebase this and ask for a retest. Then it should be good to merge.

**Comment by AndrewMeadows** [comment]
@zzmp: #10354 has merge.

**Comment by hifi-gustavo** [comment]
Build console output for commit: ca7cd61c64b4c22d022f62ddac77275a164d302b available at: https://gist.github.com/0f5176a6da4c60612886df228214d1f7<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10269-ca7cd61.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10269-ca7cd61.dmg

**Comment by claudioshg** [comment]
testing

**Comment by cozza13** [comment]
QA pass on i5

**Comment by claudioshg** [comment]
### Desktop PC - Windows 10 Home - CPU 4.0GHz Intel i7 - GTX1070 - RAM 16GB - Tested with Oculus Rift

Tested: Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10269-ca7cd61.exe

Fail:
- The interface crashes when I'm in HMD mode, I have the tablet enabled, and I close the interface from the drop-down menus. I followed the steps from the test plan, so I have 3 local injectors around my avatar. As I click to quit the interface, I'm talking, making lot's of sound, as asked to do in the test plan, and at that moment there is a crash.
hmdtabletcrash

Note: After the crash, no Splat appears.

- I've tested this PR in the same way as I mentioned above in Desktop Mode, with the Tablet and Toolbar enabled, and no crash. I've even tested Toolbar enabled in HMD and no crash. I've tested multiple times in HMD with Tablet enabled and I always get a crash.

**Comment by zzmp** [comment]
I cannot reproduce this. It sounds like a tablet crash, not an audio crash.
@claudioshg have you experienced this with any other PRs or on master?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0996feae7447fad6914674a3f3b48e9e2fb49835 available at: https://gist.github.com/0f5176a6da4c60612886df228214d1f7<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10269-0996fea.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10269-0996fea.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: defc3c1c4ff60fd7ba02091434fb133ec57ad19e available at: https://gist.github.com/0f5176a6da4c60612886df228214d1f7<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10269-defc3c1.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10269-defc3c1.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 73a3bf413bc3bdbd00cfb0f12768b0503803fa1a available at: https://gist.github.com/0f5176a6da4c60612886df228214d1f7<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10269-73a3bf4.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10269-73a3bf4.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 73a3bf413bc3bdbd00cfb0f12768b0503803fa1a available at: https://gist.github.com/0f5176a6da4c60612886df228214d1f7<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10269-73a3bf4.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10269-73a3bf4.dmg

**Comment by zzmp** [comment]
@zfox23 is trying to repro @claudioshg's crash under a debugging session. 🤞

**Comment by zfox23** [comment]
I can QA pass this.

I did notice a crash once when closing Interface from the Tablet _with a CTRL+B browser open_ after adding the three audio injectors.

However, without the CTRL+B browser open, I couldn't reproduce the same crash as @claudioshg. I'm pretty confident that this PR is not causing any new crashes.

**Pull request merged**